src/mbgl/renderer/tile_pyramid.cpp

```diff
@@ -66,6 +66,11 @@ void TilePyramid::update(const std::vector<Immutable<style::Layer::Impl>>& layer
     // If we need a relayout, abandon any cached tiles; they're now stale.
     if (needsRelayout) {
         cache.clear();
+
```

Comment:
I agree that cached and non-retained tiles need to be either discarded or updated, but I'm wondering about the specific choice you made. With these changes:

* Tiles that were already in the cache are discarded
* Tiles that are retained are updated (`setLayers` is called)
* Tiles that are not retained are updated and then go into the cache

Is that the best strategy? What makes it better than these alternatives:

* Discard both already cached and non-retained tiles
* Update both already cached and non-retained tiles

My instinct is to be pessimistic and go with the former -- assume that the cost of updating non-retained tiles is high compared to the likelihood that it'll be wasted work because the tile gets flushed from the cache before being retrieved.

Reply:
Thanks for the thoughtful review @jfirebaugh - I can see now that updating tiles before caching them is actually a bad strategy - what we should be doing instead is updating when cherry-picking them in ```createTileFn```.

Also, I liked the _discard both cached and non-retained tiles when re-layout is needed_ approach. I have a solution for that, and am now going to work on the integration tests as suggested.

The cons I can see from disposing unused tiles upon re-layout are that we might end up doing more tile creation/destruction on specific use cases - but I don't think this should be the common case.